git-protocol/src/fetch.rs

```diff
@@ -0,0 +1,493 @@
+{
+    fn handshake_extra_parameters(&self) -> Vec<(String, Option<String>)> {
+        self.opt.extra_params.clone()
+    }
+
+    fn prepare_ls_refs(
+        &mut self,
+        _: &client::Capabilities,
+        args: &mut Vec<BString>,
+        _: &mut Vec<(&str, Option<&str>)>,
+    ) -> LsRefsAction {
+        for prefix in &self.opt.ref_prefixes {
+            args.push(format!("ref-prefix {}", prefix).into())
+        }
+
+        LsRefsAction::Continue
+    }
+
+    fn prepare_fetch(
+        &mut self,
+        _: git_transport::Protocol,
+        _: &client::Capabilities,
+        _: &mut Vec<(&str, Option<&str>)>,
+        refs: &[Ref],
+    ) -> Action {
+        self.out.refs = refs.into();
+
+        Action::Continue
+    }
+
+    fn negotiate(
+        &mut self,
+        refs: &[Ref],
+        args: &mut Arguments,
+        _: Option<&Response>,
+    ) -> io::Result<Action> {
+        for r in refs {
+            if let Some(WantHave { want, have }) = (self.filter_refs)(r)? {
+                args.want(want);
+                if let Some(oid) = have {
+                    args.have(oid);
+                }
+            }
+        }
+
+        for WantRef { name, have } in &self.opt.want_refs {
+            args.want_ref(name.as_bstr());
+            if let Some(oid) = have {
+                args.have(oid);
+            }
+        }
+
+        // send done, as we don't bother with further negotiation
+        Ok(Action::Cancel)
+    }
+}
+
+#[async_trait(?Send)]
+impl<F, P, O> Delegate for Fetch<F, P, O>
+where
+    F: Fn(&Ref) -> io::Result<Option<WantHave>>,
+    P: ReceivePack<Output = O>,
+{
+    async fn receive_pack(
+        &mut self,
+        pack: impl AsyncBufRead + Unpin + 'async_trait,
+        prog: impl Progress,
+        _: &[Ref],
+        resp: &Response,
+    ) -> io::Result<()> {
+        self.out.wanted_refs = resp
+            .wanted_refs()
+            .iter()
+            .map(|response::WantedRef { id, path }| Ref::Direct {
+                path: path.clone(),
+                object: *id,
+            })
+            .collect();
+        let out = self.receive_pack.receive_pack(pack, prog).await?;
+        self.out.pack = Some(out);
+
+        Ok(())
+    }
+}
+
+pub async fn fetch<F, P, O, R, W>(
+    opt: Options,
+    filter_refs: F,
+    receive_pack: P,
+    recv: R,
+    send: W,
+) -> io::Result<Outputs<O>>
+where
+    F: Fn(&Ref) -> io::Result<Option<WantHave>> + Send + 'static,
+    P: ReceivePack<Output = O> + Send + 'static,
+    O: Send + 'static,
+    R: AsyncRead + Unpin + Send + 'static,
+    W: AsyncWrite + Unpin + Send + 'static,
+{
+    use git_transport::{
+        client::git::{ConnectMode, Connection},
+        Protocol,
+    };
+
+    let conn = Connection::new(
+        recv,
+        send,
+        Protocol::V2,
+        opt.repo.clone(),
+        None::<(String, Option<u16>)>,
+        ConnectMode::Daemon,
+    );
+    blocking::unblock(move || {
```

Comment:
It is potentially the case that this could be hidden from the user, but my understanding is this:

 the `Executor` in `blocking` is responsible for managing blocking tasks on a thread pool. In order to run async code on the thread we end up on, we need to make an “async context” available to it. Just using a thread local (async!) executor happens to work, because our `Async{Read,Write}` will receive wakeups from the outer event loop. Another way to write the same would be to use `runtime.enter()` (both `tokio` and `async-*` have equivalents). This would cause the the inner async tasks to be scheduled on the _outer_ executor.

Reply:
I would have hoped to also hear an opinion in the linked thread, but thus far there is no involvement of the maintainer.
To me it looks a good case for `git-blocking` to eventually _unblock_ `blocking`s development 😅.